### PR TITLE
Redesign the `exp`, `erf` and `erfc`  implementations

### PR DESCRIPTION
**Pull request by Thirumalai-Shaktivel**

### CODE REVIEW

**Comment by Thirumalai-Shaktivel** [comment]
Here I used a new namespace called ExpF, which is only used by LFortran.
IntrinsicScalarFunction already has Exp implemented for LPython, but it shouldn't accept complex type as input. So, I created `ExpF`

**Comment by Thirumalai-Shaktivel** [comment]
Ready!

File: src/libasr/pass/intrinsic_function_registry.h [hunk 1]
```diff
@@ -5166,12 +5191,15 @@ namespace IntrinsicScalarFunctionRegistry {
...
                  id_ == IntrinsicScalarFunctions::Trunc ||
                  id_ == IntrinsicScalarFunctions::Fix ||
                  id_ == IntrinsicScalarFunctions::Sin ||
                  id_ == IntrinsicScalarFunctions::Exp ||
                  id_ == IntrinsicScalarFunctions::Exp2 ||
                  id_ == IntrinsicScalarFunctions::Expm1 ||
+                 id_ == IntrinsicScalarFunctions::ExpF ||
                  id_ == IntrinsicScalarFunctions::Min ||
                  id_ == IntrinsicScalarFunctions::Max ||
                  id_ == IntrinsicScalarFunctions::Sqrt ||
```
Comment (certik) [inline src/libasr/pass/intrinsic_function_registry.h:5205]: Which `IntrinsicScalarFunctions` are not elemental? That is a bad design in our ASR if that is the case I think.

**Review by certik — APPROVED** [review]

File: src/libasr/pass/intrinsic_function_registry.h (hunk 1 above)
Comment (Thirumalai-Shaktivel) [reply src/libasr/pass/intrinsic_function_registry.h:5205]: How about removing this function and add a new member `elemental` to `IntrinsicScalarFunctions`?
`is_elemental` function is used in `array_op`. Do you think all the IntrinsicFunctions are elemental?

Comment (Thirumalai-Shaktivel) [reply src/libasr/pass/intrinsic_function_registry.h:5205]: I will submit a new PR to make required changes.

**Comment by Thirumalai-Shaktivel** [comment]
Thanks for the review!

File: src/libasr/pass/intrinsic_function_registry.h (hunk 1 above)
Comment (certik) [reply src/libasr/pass/intrinsic_function_registry.h:5251]: I think they are. `IntrinsicScalarFunction` should be renamed to `IntrinsicElementalFunction`:

https://github.com/lfortran/lfortran/issues/1658#issuecomment-1917519848

**Comment by certik** [comment]
Let's fix the conflicts and do one final review.

File: src/libasr/pass/intrinsic_function_registry.h [hunk 2]
```diff
@@ -54,6 +56,7 @@ enum class IntrinsicScalarFunctions : int64_t {
     Exp,
     Exp2,
     Expm1,
+    ExpF,
```
Comment (certik) [inline src/libasr/pass/intrinsic_function_registry.h:59]: What is ExpF? How is it different to Exp?

**Review by certik — APPROVED** [review]
I think that this is fine, but see my comment above.

File: src/libasr/pass/intrinsic_function_registry.h (hunk 2 above)
Comment (Thirumalai-Shaktivel) [reply src/libasr/pass/intrinsic_function_registry.h:59]: Here I used a new namespace called `ExpF`, which is only used by LFortran.
`IntrinsicScalarFunction` already has `Exp` implemented for LPython, but it shouldn't accept complex type as input. So, I created `ExpF`

Comment (Thirumalai-Shaktivel) [reply src/libasr/pass/intrinsic_function_registry.h:59]: `ExpF` -> `ExpFortran`

Comment (certik) [reply src/libasr/pass/intrinsic_function_registry.h:59]: Why cannot Exp accept complex input? It should, including in Python, I think the numpy.exp probably accepts complex input.

The LPython frontend can disallow complex input if needed, but in ASR it should be allowed.

Let's merge Exp and ExpF.

Comment (Thirumalai-Shaktivel) [reply src/libasr/pass/intrinsic_function_registry.h:59]: Got it!

**Comment by Thirumalai-Shaktivel** [comment]
Ready!

**Review by certik — APPROVED** [review]
I think that's good!

**Pull request merged**